xarray/core/variable.py

```diff
@@ -1212,35 +1212,94 @@ def __setitem__(self, key, value):
+        concat_pandas function
+        :param variables: list of variables to concatenate
+        :return: Concatenated variables
+        """
+        variable_type_set = set(map(lambda v: type(v.data), variables))
 
-        variables = list(variables)
-        first_var = variables[0]
+        if len(variable_type_set) > 1:
+            raise TypeError('Trying to concatenate variables of '
+                            'different types')
 
-        if any(not isinstance(v, cls) for v in variables):
-            raise TypeError('IndexVariable.concat requires that all input '
-                            'variables be IndexVariable objects')
+        variable_type = list(variable_type_set)[0]
+        if not variable_type == np.ndarray:
+            raise TypeError('Can only concatenate variables whose '
+                            '_data member is ndarray')
+
+        if variables[0].dtype == np.object:
+            raise TypeError('We use concat_numpy for objects whose '
+                            'dtypes are different than numpy.object ')
+
+        indexes = [v._data for v in variables]
+
+        if not indexes:
+            data = []
+        else:
+            data = np.concatenate((indexes))
+
+            if positions is not None:
+                indices = nputils.inverse_permutation(
+                    np.concatenate(positions))
+                data = data.take(indices)
+
+        return data
 
+    @classmethod
+    def concat_pandas(cls, variables, positions=None):
+        """
+        Concatenates variables. This is generic function that
+        handles all cases for which concat_numpy does not work
+        :param variables: list of variables to concatenate
+        :return: Concatenated variables
+        """
         indexes = [v._data.array for v in variables]
 
         if not indexes:
             data = []
         else:
+
             data = indexes[0].append(indexes[1:])
 
             if positions is not None:
                 indices = nputils.inverse_permutation(
                     np.concatenate(positions))
                 data = data.take(indices)
 
+        return data
+
+    @classmethod
+    def concat(cls, variables, dim='concat_dim', positions=None,
+               shortcut=False):
+        """Specialized version of Variable.concat for
+        IndexVariable objects.
+        This exists because we want to avoid converting Index objects to NumPy
+        arrays, if possible.
+        """
+        if not isinstance(dim, basestring):
+            dim, = dim.dims
+
+        variables = list(variables)
+
+        first_var = variables[0]
+
+        if any(not isinstance(v, cls) for v in variables):
+            raise TypeError('IndexVariable.concat requires that all input '
+                            'variables be IndexVariable objects')
+
+        # GH1434
+        # Fixes bug: "xr.concat loses coordinate dtype
+        # information with recarrays in 0.9"
+        try:
```

Comment:
Rather than using exceptions for control flow, can we divide this into two cases based upon whether any (all?) of the variables have dtype=object?